File: backend/app/schemas/node.py
from pydantic import BaseModel
from datetime import datetime
from uuid import UUID
from typing import Optional, Dict, Any, List
from enum import Enum


class NodeType(str, Enum):
    IMAGE = "image"
    PROMPT = "prompt"
    VIDEO = "video"
    CONTAINER = "container"
    RATIO = "ratio"
    SCENE = "scene"
    EXTENSION = "extension"
    CHARACTER = "character"
    PRODUCT = "product"
    SETTING = "setting"


class NodeStatus(str, Enum):
    IDLE = "idle"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


class NodeCreate(BaseModel):
    type: NodeType
    position_x: float = 0
    position_y: float = 0
    data: Dict[str, Any] = {}
    character_id: Optional[UUID] = None


class NodeUpdate(BaseModel):
    position_x: Optional[float] = None
    position_y: Optional[float] = None
    data: Optional[Dict[str, Any]] = None
    character_id: Optional[UUID] = None


class NodeResponse(BaseModel):
    id: UUID
    project_id: UUID
    character_id: Optional[UUID] = None
    type: NodeType
    position_x: float
    position_y: float
    data: Dict[str, Any]
    status: NodeStatus
    error_message: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class NodeBatchUpdate(BaseModel):
    nodes: List[NodeUpdate]


class BranchConfig(BaseModel):
    offset_y: float = 250
    modifications: Optional[Dict[str, Dict[str, Any]]] = None
    # e.g., {"ref_id_of_branch_point": {"script_text": "Different hook..."}}


class ConnectionResponse(BaseModel):
    id: UUID
    project_id: UUID
    source_node_id: UUID
    target_node_id: UUID
    source_handle: Optional[str] = None
    target_handle: Optional[str] = None

    class Config:
        from_attributes = True


class BranchResponse(BaseModel):
    cloned_nodes: List[NodeResponse]
    cloned_connections: List[ConnectionResponse]
    branch_group_id: str

File: frontend/lib/types/node.ts
// Node editor types

export type NodeType = 'image' | 'prompt' | 'video' | 'container' | 'ratio' | 'scene' | 'extension' | 'character' | 'product' | 'setting' | 'stitch';
export type NodeStatus = 'idle' | 'processing' | 'completed' | 'failed';

export interface Node {
  id: string;
  project_id: string;
  type: NodeType;
  position_x: number;
  position_y: number;
  data: Record<string, any>;
  status: NodeStatus;
  character_id?: string;
  error_message?: string;
  created_at: string;
  updated_at: string;
}

export interface Connection {
  id: string;
  project_id: string;
  source_node_id: string;
  target_node_id: string;
  source_handle?: string;
  target_handle?: string;
  created_at: string;
}

export interface NodeHandle {
  id: string;
  type: 'input' | 'output';
  position: { x: number; y: number };
}
